File: solutions/clojure/156-decode-string.clj
;; Problem 156: Decode String (LeetCode 394)
;; Difficulty: Med
;; Language: Clojure
;; 
(defn decode-string [s]
  (letfn [(helper [chars]
            (loop [cs chars acc ""]
              (cond
                (or (empty? cs) (= (first cs) \])) [acc (rest cs)]
                (Character/isDigit (first cs))
                  (let [[num-chars rest] (split-with #(Character/isDigit %) cs)
                        n (Integer/parseInt (apply str num-chars))
                        [inner remaining] (helper (rest rest))
                        [next final] (helper remaining)]
                    [(str acc (apply str (repeat n inner)) next) final])
                :else (recur (rest cs) (str acc (first cs))))))]
    (first (helper (seq s)))))

File: solutions/scala/156-decode-string.scala
/**
 * Problem 156: Decode String (LeetCode 394)
 * Difficulty: Med
 * Language: Scala
 */
def decodeString(s: String): String = {
  def helper(i: Int): (String, Int) = {
    val sb = new StringBuilder; var idx = i
    while (idx < s.length && s(idx) != ']') {
      if (s(idx).isDigit) {
        var num = 0
        while (s(idx).isDigit) { num = num * 10 + (s(idx) - '0'); idx += 1 }
        idx += 1 // skip '['
        val (decoded, ni) = helper(idx)
        sb.append(decoded * num); idx = ni + 1
      } else { sb.append(s(idx)); idx += 1 }
    }
    (sb.toString, idx)
  }
  helper(0)._1
}
